# Spotify Analysis

First we install the necessary libraries needed for data visualization and wrangling. Additionally, the R wrapper, Spotifyr, which allows us to get insights from Spotify's web api is installed as well. 

In [33]:
library(tidyverse)
library(spotifyr)
library(repr)

Below, we access the web API using the Client ID and Client Secret provided by Spotify upon creating a Developer Account.

In [34]:
access_token <- get_spotify_access_token()

## A Comparison Of My Playlists From Year 1 and Year 2 Of Undergrad
I have attempted to capture each year of my undergraduate degree in a unique playlist. So far, I have one and a half years' worth of data. I will begin my analysis using a playlist I began creating in the fall of 2022, my first year of university. I am interested in learning the general atmosphere of the playlist and how my music preference has changed throughout my first and second years of university. 

### Data Loading & Cleaning

> While reading my playlist data, I realized that Spotify has a limit that only allows 100 items to be read at a time. Therefore, to add all 566 items of my playlist here, I set the offsets parameter to increment by 100 until it reaches 500. Then I map each of these offsets to the data frames of 100 tracks, starting from each offset value. Finally I combine all the separate data frames into a single data frame. 

In [57]:
# Loading the playlist data frame into the "playlist_2022" variable

playlist_id <- "2cQLZcVSOLDfY03SY5DZrf"
offsets <- seq(0, 500, 100)

# Mapping offsets
playlist_tracks <- map(offsets, ~ get_playlist_tracks(playlist_id, limit = 100, offset = .x))

# Combining the results into a single data frame
y1_playlist <- bind_rows(playlist_tracks) |> as_tibble()


In [58]:
# Getting all the column names of the data frame
colnames(y1_playlist)

[1] "added_at"                           "is_local"                          
 [3] "primary_color"                      "added_by.href"                     
 [5] "added_by.id"                        "added_by.type"                     
 [7] "added_by.uri"                       "added_by.external_urls.spotify"    
 [9] "track.artists"                      "track.available_markets"           
[11] "track.disc_number"                  "track.duration_ms"                 
[13] "track.episode"                      "track.explicit"                    
[15] "track.href"                         "track.id"                          
[17] "track.is_local"                     "track.name"                        
[19] "track.popularity"                   "track.preview_url"                 
[21] "track.track"                        "track.track_number"                
[23] "track.type"                         "track.uri"                         
[25] "track.album.album_type"             "track.album.artists"               
[27] "track.album.available_markets"      "track.album.href"                  
[29] "track.album.id"                     "track.album.images"                
[31] "track.album.name"                   "track.album.release_date"          
[33] "track.album.release_date_precision" "track.album.total_tracks"          
[35] "track.album.type"                   "track.album.uri"                   
[37] "track.album.external_urls.spotify"  "track.external_ids.isrc"           
[39] "track.external_urls.spotify"        "video_thumbnail.url"

Based on the above columns, some interesting columns to analyze include: 
- 'added_at' - The date and time the song was added (NA is used if playlist is too old)
- 'track.artists' - 
- 'track.duration_ms'
- 'track.explicit'
- 'track.name'
- 'track.popularity'
- 'track.album.album_type'
- 'track.album.name'
- 'track.album.release_date'

In [81]:
# Selecting only the identified columns above
shortened_y1_playlist <- y1_playlist |> select(added_at, track.artists, track.duration_ms, 
                                                  track.explicit, track.name, track.popularity, 
                                                  track.album.album_type, track.album.name, 
                                                  track.album.release_date)
# Getting the dimensions
dim(shortened_y1_playlist)

# Getting the first six rows of our shorter data frame
head(shortened_y1_playlist)

[1] 567   9

added_at            
1 2022-08-06T00:01:24Z
2 2022-08-06T00:04:37Z
3 2022-08-06T00:17:48Z
4 2022-08-06T00:18:07Z
5 2022-08-06T00:26:55Z
6 2022-08-10T04:30:57Z
  track.artists                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
1 https://api.spotify.com/v1/artists/5GGJosGMs08YEmKTZJe1fL, 5GGJosGMs08YEmKTZJe1fL                                   , Suki Waterhouse                                          , artist                                                   , spotify:artist:5GGJosGMs08YEmKTZJe1fL                    , https://open.spotify.com/artist/5GGJosGMs08YEmKTZJe1fL                                                                                                                                                                                                                                                                                                                                                                     
2 https://api.spotify.com/v1/artists/5ZsFI1h6hIdQRw2ti0hz81, 5ZsFI1h6hIdQRw2ti0hz81                                   , ZAYN                                                     , artist                                                   , spotify:artist:5ZsFI1h6hIdQRw2ti0hz81                    , https://open.spotify.com/artist/5ZsFI1h6hIdQRw2ti0hz81                                                                                                                                                                                                                                                                                                                                                                     
3 https://api.spotify.com/v1/artists/2kxP07DLgs4xlWz8YHlvfh, 2kxP07DLgs4xlWz8YHlvfh                                   , NIKI                                                     , artist                                                   , spotify:artist:2kxP07DLgs4xlWz8YHlvfh                    , https://open.spotify.com/artist/2kxP07DLgs4xlWz8YHlvfh                                                                                                                                                                                                                                                                                                                                                                     
4 https://api.spotify.com/v1/artists/6nxWCVXbOlEVRexSbLsTer, https://api.spotify.com/v1/artists/6xHUXzrfhFgnIv86EBR3Ml, 6nxWCVXbOlEVRexSbLsTer                                   , 6xHUXzrfhFgnIv86EBR3Ml                                   , Flume                                                    , kai                                                      , artist                                                   , artist                                                   , spotify:artist:6nxWCVXbOlEVRexSbLsTer                    , spotify:artist:6xHUXzrfhFgnIv86EBR3Ml                    , https://open.spotify.com/artist/6nxWCVXbOlEVRexSbLsTer   , https://open.spotify.com/artist/6xHUXzrfhFgnIv86EBR3Ml   
5 https://api.spotify.com/v1/artists/3pc0bOVB5whxmD50W79wwO, 3pc0bOVB5whxmD50W79wwO                                   , keshi                                                    , artist                                                   , spotify:artist:3pc0bOVB5whxmD50W79wwO                    ,

According to the dimensions of the data frame, there are 567 rows. However, my playlist only has 566 songs! This means there could be some overcounting involved. Let's check if there are duplicate rows:

In [60]:
# Keeping only unique rows from shortened_playlist_2022 and checking if total rows become 566
dim(distinct(shortened_y1_playlist))

[1] 567   9

This shows us that there were no overcounted songs, since even after removing duplicates we are left with 567 rows. To identify the source of the extra row, let's check if there are any rows with NA's:

In [61]:
rows_with_missing <- shortened_y1_playlist[apply(is.na(shortened_y1_playlist), 1, any), ]
rows_with_missing

added_at             track.artists track.duration_ms track.explicit
1 2023-07-30T20:14:01Z NULL          NA                NA            
  track.name track.popularity track.album.album_type track.album.name
1 NA         NA               NA                     NA              
  track.album.release_date
1 NA

Above is the row that caused the extra row count in our playlist data frame. Perhaps Spotify puts NA for songs that have been deleted from their platform. Since there is no data in the row, we can safely remove this row from our playlist data frame:

In [62]:
shortened_playlist_2022 <- na.omit(shortened_y1_playlist)
dim(shortened_y1_playlist)

[1] 567   9

I will now also load my year 2 playlist which has 102 songs in it:

In [79]:
playlist_id <- "6paPUoyazZ2SkgZ5HkYpoW"
offsets <- c(0, 100)
segmented_playlist_tracks <- map(offsets,  ~ get_playlist_tracks(playlist_id, limit = 100, offset = .x))
shortened_y2_playlist <- bind_rows(segmented_playlist_tracks) |>
                         as_tibble() |>
                         select(added_at, track.artists, track.duration_ms, 
                                          track.explicit, track.name, track.popularity, 
                                          track.album.album_type, track.album.name, 
                                          track.album.release_date)

# Getting dimensions and last 6 rows of the y2 playlist
dim(shortened_y2_playlist)
tail(shortened_y2_playlist)


[1] 101   9

added_at            
1 2023-12-10T18:03:34Z
2 2023-12-10T18:03:48Z
3 2023-12-15T13:29:11Z
4 2023-12-21T15:40:15Z
5 2023-12-22T03:36:04Z
6 2024-01-03T00:22:13Z
  track.artists                                                                                                                                                                                                                                                                                                                                                   
1 https://api.spotify.com/v1/artists/6eJa3zG1QZLRB3xgRuyxbm, 6eJa3zG1QZLRB3xgRuyxbm                                   , Dayglow                                                  , artist                                                   , spotify:artist:6eJa3zG1QZLRB3xgRuyxbm                    , https://open.spotify.com/artist/6eJa3zG1QZLRB3xgRuyxbm   
2 https://api.spotify.com/v1/artists/62iEYJuqOpo6M5UAu4aw3U, 62iEYJuqOpo6M5UAu4aw3U                                   , dad sports                                               , artist                                                   , spotify:artist:62iEYJuqOpo6M5UAu4aw3U                    , https://open.spotify.com/artist/62iEYJuqOpo6M5UAu4aw3U   
3 https://api.spotify.com/v1/artists/4gzpq5DPGxSnKTe4SA8HAU, 4gzpq5DPGxSnKTe4SA8HAU                                   , Coldplay                                                 , artist                                                   , spotify:artist:4gzpq5DPGxSnKTe4SA8HAU                    , https://open.spotify.com/artist/4gzpq5DPGxSnKTe4SA8HAU   
4 https://api.spotify.com/v1/artists/3g65HWtAf8uyVmpR4sRbrT, 3g65HWtAf8uyVmpR4sRbrT                                   , The Hails                                                , artist                                                   , spotify:artist:3g65HWtAf8uyVmpR4sRbrT                    , https://open.spotify.com/artist/3g65HWtAf8uyVmpR4sRbrT   
5 https://api.spotify.com/v1/artists/5XQRfs0gXs30aWUn1Umves, 5XQRfs0gXs30aWUn1Umves                                   , Choker                                                   , artist                                                   , spotify:artist:5XQRfs0gXs30aWUn1Umves                    , https://open.spotify.com/artist/5XQRfs0gXs30aWUn1Umves   
6 https://api.spotify.com/v1/artists/3zmfs9cQwzJl575W1ZYXeT, 3zmfs9cQwzJl575W1ZYXeT                                   , Men I Trust                                              , artist                                                   , spotify:artist:3zmfs9cQwzJl575W1ZYXeT                    , https://open.spotify.com/artist/3zmfs9cQwzJl575W1ZYXeT   
  track.duration_ms track.explicit track.name          track.popularity
1 215555            FALSE          Listerine           59              
2 201857            FALSE          out 4 a breather    46              
3 278906             TRUE          Trouble In Town     49              
4 200961            FALSE          Fiona               41              
5 231680             TRUE          Juno                52              
6 205554            FALSE          I Hope to Be Around 59              
  track.album.album_type track.album.name                  
1 album                  Fuzzybrain                        
2 single                 I AM JUST A BOY LEAVE ME ALONE !!!
3 album                  Everyday Life                     
4 album                  What's Your Motive                
5 album                  Honeybloom                        
6 single                 I Hope to Be Around               
  track.album.release_date
1 2019-11-14              
2 2021-03-19              
3 2019-11-22              
4 2023-09-08              
5 2018-08-03              
6 2017-11-10

### A Visual Analysis